src/link/link_info/geneve.rs

```diff
@@ -0,0 +1,221 @@
+use byteorder::{BigEndian, ByteOrder, NativeEndian};
+use netlink_packet_utils::{
+    nla::{DefaultNla, Nla, NlaBuffer},
+    parsers::{parse_u16_be, parse_u32, parse_u32_be, parse_u8},
+    traits::Parseable,
+    DecodeError,
+};
+
+const IFLA_GENEVE_ID: u16 = 1;
+const IFLA_GENEVE_REMOTE: u16 = 2;
+const IFLA_GENEVE_TTL: u16 = 3;
+const IFLA_GENEVE_TOS: u16 = 4;
+const IFLA_GENEVE_PORT: u16 = 5;
+const IFLA_GENEVE_COLLECT_METADATA: u16 = 6;
+const IFLA_GENEVE_REMOTE6: u16 = 7;
+const IFLA_GENEVE_UDP_CSUM: u16 = 8;
+const IFLA_GENEVE_UDP_ZERO_CSUM6_TX: u16 = 9;
+const IFLA_GENEVE_UDP_ZERO_CSUM6_RX: u16 = 10;
+const IFLA_GENEVE_LABEL: u16 = 11;
+const IFLA_GENEVE_TTL_INHERIT: u16 = 12;
+const IFLA_GENEVE_DF: u16 = 13;
+const IFLA_GENEVE_INNER_PROTO_INHERIT: u16 = 14;
+
+const GENEVE_DF_UNSET: u8 = 0;
+const GENEVE_DF_SET: u8 = 1;
+const GENEVE_DF_INHERIT: u8 = 2;
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[non_exhaustive]
+pub enum GeneveDf {
+    Unset,
+    Set,
+    Inherit,
+    Other(u8),
+}
+
+impl From<u8> for GeneveDf {
+    fn from(d: u8) -> Self {
+        match d {
+            GENEVE_DF_UNSET => Self::Unset,
+            GENEVE_DF_SET => Self::Set,
+            GENEVE_DF_INHERIT => Self::Inherit,
+            _ => Self::Other(d),
+        }
+    }
+}
+
+impl From<GeneveDf> for u8 {
+    fn from(d: GeneveDf) -> Self {
+        match d {
+            GeneveDf::Unset => GENEVE_DF_UNSET,
+            GeneveDf::Set => GENEVE_DF_SET,
+            GeneveDf::Inherit => GENEVE_DF_INHERIT,
+            GeneveDf::Other(value) => value,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[non_exhaustive]
+pub enum InfoGeneve {
+    Id(u32),
+    Remote(Ipv4Addr),
+    Remote6(Ipv6Addr),
+    Ttl(u8),
+    Tos(u8),
+    Port(u16),
+    CollectMetadata(bool),
+    UdpCsum(bool),
+    UdpZeroCsum6Tx(bool),
+    UdpZeroCsum6Rx(bool),
+    Label(u32),
+    TtlInherit(bool),
+    Df(GeneveDf),
+    InnerProtoInherit(bool),
+    Other(DefaultNla),
+}
+
+impl Nla for InfoGeneve {
+    fn value_len(&self) -> usize {
+        use self::InfoGeneve::*;
+        match self {
+            Id(_) | Remote(_) | Label(_) => 4,
+            Remote6(_) => 16,
+            Ttl(_) | Tos(_) | UdpCsum(_) | UdpZeroCsum6Tx(_)
+            | UdpZeroCsum6Rx(_) | TtlInherit(_) | Df(_) => 1,
+            Port(_) => 2,
+            CollectMetadata(_) | InnerProtoInherit(_) => 0,
+            Other(nla) => nla.value_len(),
+        }
+    }
+
+    fn emit_value(&self, buffer: &mut [u8]) {
+        use self::InfoGeneve::*;
+        match self {
+            Id(value) => NativeEndian::write_u32(buffer, *value),
+            Remote(value) => buffer.copy_from_slice(&value.octets()),
+            Remote6(value) => buffer.copy_from_slice(&value.octets()),
+            Ttl(value) | Tos(value) => buffer[0] = *value,
+            Port(value) => BigEndian::write_u16(buffer, *value),
+            CollectMetadata(_value) | InnerProtoInherit(_value) => (),
```

Comment:
For `nla_put_flag()`, we should just use `Foo` because the existence of it means true, otherwise we are incorrectly emitting value for `Foo(false)`.

Reply:
Changed.